Applications/Spire/Include/Spire/KeyBindings/KeyBindings.hpp

```diff
@@ -220,7 +219,13 @@ namespace Details {
 
     private:
       using Actions = Nexus::RegionMap<boost::optional<Action>>;
-      QHash<QKeySequence, Actions> m_bindings;
```

Comment:
It was changed so that the key bindings could be retrieved in the order they were set.

Reply:
Okay, can you identify the exact actions / API calls that require the original order? We may be able to come up with a better solution.

Reply:
It was a requirement for the Task Keys table (order bindings table): that the user's ordering of their bindings is preserved when the Key Bindings window is closed and reopened.

Reply:
Let me put it this way: would it be enough if only `build_order_bindings` and `build_action_bindings` methods returned the vectors sorted by the order of the binding?

Reply:
As far as I'm aware, yes, it would be enough for those methods to return the vectors sorted by insertion order.

Reply:
Ok, now it seems like a pattern used in Python OrderedDict and stuff like that. There are different ways to implement that better than using a plain array, but since you are only concerned about returning the entire collection in the insertion order, I recommend the following:
1. Recover QHash and leave store, remove, and find operations intact.
2. Add a sequence number to the actions stored in the hash table, and increment that counter inside store operations.
3. Inside build_- functions, return the collection sorted by the sequence number. Of course, don't expose the sequence numbers and remove them from the output.

This solution may not be optimal in terms of strict computational complexity, but it is quite local (you needed yo change the build- family of functions, so you mostly modified it and not the rest of the code) and simple. If those build- methods end up getting called very frequently, you can easily achieve linear complexity by adding a bit more code and complexity to store and remove methods.

Reply:
I just realized that it might be not more difficult to implement the optimal solution:
In the above, instead of adding sequence numbers to the hash table values, create an `std::map` of sequence numbers to action bindings. Then, on each store and delete operation, do the insertion/removal on both the hash table and sequence map. The find operation remains intact, while build will use the `std::map` to produce the output in linear time.
You can figure out the details like whether to duplicate action records or store pointers to either structure yourself.

Reply:
This may be significantly more complicated than it needs to be. Let's stick to the straight forward implementation for now and if there ends up being some critical performance issue then the implementation can be changed in the future, but really given that there are like 20 key bindings it's overkill to implement this solution.

Reply:
The rationale for the above is not just performance of some UI component, but it mostly has to do with problem solving, code re-use, readability, coding style, this kind of thing. Going to std::vector required lots of code changes, and there is a reason why STL does not provide shorthand syntax for those operations.
When a new requirement arrives, we try to satisfy it with the fewest changes possible to the behavior and code of the existing parts. Strictly speaking, in order for us to introduce an order to the list output, all we need is to store the elements in `std::list`, and make the hash table elements into list iterators. Only minor indirection (replace "." with "->") is introduced to all existing operations as a result. That is also among the most natural solutions to the problem when we think about adding a new feature, not re-implementing the whole thing.

Anyway, if you think it is not worth it, let it be as is.